### ISSUE TITLE
Bi-directional dependency exception when attempting to create singleton that had previously thrown an exception during construction

### ISSUE THREAD

**Issue by morstaine** [structuremap/structuremap#329]
When attempting to build a singleton, either directly or indirectly as a dependency of something else, in cases where building that singleton throws an exception (database down, network issue, just a plain old bug, etc), the first time it is constructed it will fail with an appropriate exception. However, on subsequent attempts to build the same singleton, you will repeatedly get the "Bi-directional dependency relationship detected!" error, even when no such relationship exists.

This makes it difficult to trace the original cause of the exception.

I would expect that subsequent attempts to construct the singleton should either succeed if the root cause for the exception was corrected, or throw a meaningful exception if the root cause had not been corrected.

Attached is a failing test case I would expect to work:

```csharp
using System;
using Microsoft.VisualStudio.TestTools.UnitTesting;
using StructureMap;

namespace ReproduceStructureMapBiDiProblem
{
    [TestClass]
    public class ReproduceBiDiProblem
    {
        [TestMethod]
        public void Singleton_WhenFailsToConstruct_ShouldNotThrowBiDiErrorsOnFurtherAttemptsToConstruct()
        {
            var container = new Container(_ => 
                _.ForSingletonOf<SomeSingletonThatConnectsToDatabaseOnConstruction>()
                    .Use<SomeSingletonThatConnectsToDatabaseOnConstruction>());

            // 1st attempt : database is down, expect "Database is down" exception
            DatabaseStatus.DatabaseIsDown = true;

            try
            {
                // If the database is down, the construction of the instance will fail.
                var attempt1 = container.GetInstance<SomeSingletonThatConnectsToDatabaseOnConstruction>();

                Assert.Fail("Expected container.GetInstance to throw Exception");
            }
            catch (Exception ex)
            {
                // The first time we attempt, we get a valuable error
                Assert.IsTrue(ex.ToString().Contains("Database is down"));
            }

            // 2nd attempt : database is still down, still expect "Database is down" exception
            try
            {
                // If the database is down, the construction of the instance will fail.
                var attempt2 = container.GetInstance<SomeSingletonThatConnectsToDatabaseOnConstruction>();

                Assert.Fail("Expected container.GetInstance to throw Exception");
            }
            catch (Exception ex)
            {
                // The second time we attempt, we _should_ get a valuable error, but get a bi-di error

                // This assert currently fails in SM 3.x
                Assert.IsTrue(ex.ToString().Contains("Database is down"));
            }

            // The database has now come online.
            DatabaseStatus.DatabaseIsDown = false;

            // If the database is up, the construction of the instance should now succeed, but
            // instead still throws a bi-di error
            var attempt3 = container.GetInstance<SomeSingletonThatConnectsToDatabaseOnConstruction>();

            Assert.IsNotNull(attempt3);
        }
    }

    public static class DatabaseStatus
    {
        public static bool DatabaseIsDown { get; set; }
    }

    public class SomeSingletonThatConnectsToDatabaseOnConstruction
    {
        public SomeSingletonThatConnectsToDatabaseOnConstruction()
        {
            if (DatabaseStatus.DatabaseIsDown)
                throw new Exception("Database is down");
        }
    }
}

**Comment by jeremydmiller** [comment]
@morstaine Got it. No promises, but I'll try to look at this some night this week or the weekend. Issues w/ failing tests go to the front of the line every time though;-)

**Comment by girtsl** [comment]
I believe I'm running into the same issue. At first I thought it was issue fixed with this [PR](https://github.com/structuremap/structuremap/pull/271) as it doesn't happen very often - about once a week or even rarer - and it's a false positive, but digging deeper revealed the correlation with intermittent network connectivity issues.
I'm using the latest version available on nuget - 3.1.4.143.

**Comment by jeremydmiller** [comment]
I'm finally looking into this right now
